File: business_logic/src/main/java/org/metplus/curriculum/domain/job/CreateJob.java
package org.metplus.curriculum.domain.job;

import org.metplus.curriculum.domain.Job;

public class CreateJob<T> {

    private JobRepository jobRepository;

    public CreateJob(JobRepository jobRepository) {
        this.jobRepository = jobRepository;
    }

    public T create(String jobId, String title, String description, ResultHandler<T> resultHandler) {
        Job existingJob = jobRepository.findById(jobId);
        if (existingJob == null) {
            Job newJob = Job.builder()
                    .id(jobId)
                    .title(title)
                    .description(description)
                    .build();
            try {
                jobRepository.save(newJob);
            } catch (Exception exp) {

            }
            return resultHandler.success();
        } else {
            return resultHandler.jobExists(jobId);
        }
    }

    public interface ResultHandler<T> {
        T jobExists(String jobId);
        T fatalError();
        T success();
    }
}

File: business_logic/src/main/java/org/metplus/curriculum/domain/job/CreateJob.kt
package org.metplus.curriculum.domain.job

import org.metplus.curriculum.domain.Job

class CreateJob(private val jobRepository: JobRepository) {

    fun <T> create(jobId: String, title: String, description: String, resultHandler: ResultHandler<T>): T {
        val existingJob = jobRepository.findById(jobId)
        if (existingJob == null) {
            val newJob = Job(
                    jobId = jobId,
                    title = title,
                    description = description
            )
            try {
                jobRepository.save(newJob)
            } catch (exp: Exception) {
                return resultHandler.fatalError(jobId)
            }
            return resultHandler.success()
        } else {
            return resultHandler.jobExists(jobId)
        }
    }

    interface ResultHandler<T> {
        fun jobExists(jobId: String): T
        fun fatalError(jobId: String): T
        fun success(): T
    }
}
